### PR TITLE
Reuse existing season items across library buckets on completion

### LINKED ISSUE #812: TV completion can fork imported season identities and infer completion from sparse history
## Summary

Completing a whole TV show can create a second season identity when an imported season already exists in another compatible library bucket. The duplicate can split the user's episode history and dates across two season rows.

The same completion path can also infer `Completed` from the highest episode number instead of the number of distinct completed episodes when release events are unavailable. For example, watching only episode 10 of a ten-episode season must not complete the season.

## Reported impact

PR #760 reports an imported *Breaking Bad* library where seasons in the `tv` bucket held the user's April 2025 dates. Completing the show created new season-bucket rows and a second set of 62 episodes dated on the completion day. The original dates remained attached to the first rows.

## Reproduction shape

1. Import a TV show, its seasons, and episode history so each season `Item` uses the show's `tv` bucket.
2. Keep the original episode dates on those rows.
3. Mark the whole TV show as completed.
4. `TV._completed()` resolves only the canonical completion bucket and creates a parallel `Item`/`Season`/episode set instead of reusing the compatible imported season.
5. In a season with no release-event rows, set `local_season_episode_count=10` and record only episode 10. A maximum-position comparison can mark the season completed even though nine episodes have no completed evidence.

## Expected behavior

- Reuse an existing season item only from a bucket that is compatible with the parent TV identity.
- Never attach a normal TV completion to an anime-bucket season, or an anime completion to an unrelated TV identity.
- Preserve existing `Season`, `Episode`, start-date, end-date, score, note, and history rows.
- Use an atomic create path when no compatible season item exists.
- Derive local completion from the count of distinct completed episodes, not the highest episode number.
- Preserve a deliberate `In progress` status used for a rewatch.

## Acceptance criteria

- [ ] A TV-bucket imported season is reused when the matching TV show is completed.
- [ ] No second season `Item`, `Season`, or episode set is created.
- [ ] Existing watch dates remain attached and visible.
- [ ] Normal TV and anime identity buckets cannot cross-attach.
- [ ] Concurrent completion attempts converge without an avoidable create race or duplicate-key error.
- [ ] One sparse late-episode watch does not complete the season.
- [ ] All distinct completed episodes can complete a season when `local_season_episode_count` is authoritative.
- [ ] A manual `In progress` rewatch remains `In progress`.
- [ ] Focused model tests cover each case.
- [ ] `uv run --no-sync ruff check src` and the affected model tests pass.

## Relationships

- Proposed implementation: #760.
- Related identity work: #544, #620, #623, and #639.

## Notes for contributors

The objective in #760 is valid. The implementation needs an allowed-bucket rule, the existing database-safe create behavior, and count-based completion evidence before it is safe to merge. This issue keeps those requirements in one place for the current contributor and future fixes.

### PR DESCRIPTION
**Pull request by JoshDaFishy**
## Summary

Complete a TV show without creating a second season identity when a compatible imported season already exists in another library bucket. Use local episode-count evidence when provider release events are unavailable.

This objective is valid. The current draft needs the corrections in the review section before it is safe to merge.

## User-facing problem

`TV._completed()` creates or resolves season `Item` rows while it expands a completed show into seasons and episodes. Imported seasons can use the parent show's `tv` bucket, while this path historically looked in the `season` bucket. The mismatch can create a second season and episode set. Existing dates remain on the first set, while the newly generated set receives completion dates from the later action.

The draft also uses the highest watched episode number as proof that all episodes were completed when release-event rows are absent. Sparse history makes that unsafe. Watching only episode 10 of a ten-episode season is one completed episode, not ten.

## Reported evidence

The contributor observed imported *Breaking Bad* seasons in the `tv` bucket with April 2025 dates. Completing the show created another set in the `season` bucket and duplicated 62 episodes with dates from the completion action.

## Proposed solution

1. Resolve an existing season only from buckets that are compatible with the parent TV identity.
2. Prefer the parent show's bucket, then the canonical season bucket for normal TV.
3. Do not allow normal TV and anime identity buckets to cross-attach.
4. If no compatible row exists, retain `get_or_create()` for the intended target bucket.
5. When `local_season_episode_count` is authoritative, compare it with the count of distinct completed episodes, not the highest episode position.
6. Preserve a deliberate `In progress` status used for a rewatch.

### 1. Sparse history can be marked complete

The new `max_watched >= local_total` condition is not completion evidence. Existing model behavior deliberately separates progress position from distinct completed-episode count. A season with only episode 10 watched has a progress position of 10 but a completed count of 1.

Use `self.completed_episode_count >= local_total` for this fallback and add tests for sparse history, duplicate plays, full distinct completion, and the manual rewatch override.

### 2. The create path is no longer database-safe

The draft changes an atomic `get_or_create()` path into a lookup followed by `Item.objects.create()`. Two completion requests can both see no row and then race to create the same target identity. One request can fail with a duplicate-key error.

Keep the cross-bucket lookup, but use `get_or_create()` when the target bucket is absent.

### 3. The fallback lookup is too broad for TV/anime identity

`find_item_across_buckets()` returns the oldest candidate from any bucket when the preferred bucket is absent. In this model path, that can select an anime-bucket season for a normal TV parent, or the reverse. Existing identity work in #623 shows why those parent identities must stay separate.

Restrict the candidate search to an ordered set of allowed buckets. For a normal TV parent, prefer the parent `tv` bucket and then `season`. For an anime parent, use the anime identity only unless a separate, tested migration rule says otherwise.

### 4. Regression tests are required

This PR changes persisted item selection, season relationships, episode history visibility, and status reconciliation. It currently adds no tests and records no command results.

Required cases are tracked in #812.

### 5. Update from current `latest`

The branch is 46 commits behind and 1 commit ahead of current `latest` as reviewed on 2026-08-16. Current model and importer changes include distinct-episode completion behavior that this PR must preserve. Update the branch before final validation.

## AI assistance

Provider: Anthropic. Model: Claude Opus 5 (API identifier claude-opus-5).

## Validation

No executable validation result is recorded on the current branch. GitHub Actions reports `action_required`; App Tests, Lint, Docker Image, and CodeQL have not executed.

Run and record at minimum after updating from `latest`:

- `scripts/test.sh app.tests.models.test_tv_completed_on_create app.tests.models.test_season`
- the new focused regression-test module for #812
- `uv run --no-sync ruff check src`
- `uv run --no-sync python src/manage.py makemigrations --check --dry-run`
- `scripts/test.sh`
- `git diff --check upstream/latest...HEAD`

Do not mark a command complete unless its result is available from the updated branch.

## Contract handoff

- Domain guide regeneration: Not applicable unless the implementation changes domain vocabulary.
- OpenAPI regeneration: Not applicable. No API contract is proposed.
- Contract tests: Record the result if a shared contract or serializer changes. No such change exists in the current draft.
- Migration: No schema change is proposed. `makemigrations --check --dry-run` must confirm this.

## Screenshots and interaction QA

Not applicable for the current scope. This is backend model behavior. It does not change templates, CSS, layout, keyboard behavior, screen-reader output, focus order, visual grouping, or cognitive load.

The user-facing result still needs a manual data check: existing dates and history must remain visible after completing the parent show.

## Human review

- [x] Completed on 2026-08-16.
- [x] Objective confirmed as useful.
- [x] Three blocking correctness findings recorded.
- [ ] Corrective commit reviewed.
- [ ] Updated-branch test evidence reviewed.

## `/qa`

- [x] Persisted-identity path reviewed.
- [x] Status-transition path reviewed.
- [x] Sparse-history and repeat-watch behavior reviewed.
- [x] TV/anime identity boundary reviewed.
- [x] Current-target divergence reviewed.
- [ ] Executable QA pending corrective commit and branch update.

## Post-mortem

**Trigger:** An imported season and the whole-show completion path used different library buckets for the same provider identity.

**Failure:** Completion did not find the compatible imported row and created a parallel season/episode set. A second fallback then treated the highest watched episode number as a completed-episode count.

**Impact:** Users can see duplicated seasons or episodes, split date history, and an incorrect completed status.

**Why the defect escaped:** Tests did not combine imported bucket identity, whole-show completion, existing watch dates, sparse episode history, and anime/TV identity separation.

**Corrective action:** Use an ordered compatible-bucket lookup, retain the atomic create path, derive completion from distinct completed episodes, and add focused model tests.

**Prevention:** Keep the #812 acceptance matrix as the regression contract for future import and completion changes.

## Relationships

- Fixes #812 when all acceptance criteria are met.
- Refs #544 for bucket-aware item identity.
- Refs #620 for duplicate TV/season identities.
- Refs #623 for TV/anime parent-identity separation.
- Refs #639 for duplicate episode effects.

### CODE REVIEW

File: src/app/models/tv.py [hunk 1]
```diff
@@ -349,24 +349,40 @@ def _completed(self):
...
+                    ),
                     **Item.title_fields_from_metadata(
                         season_metadata,
                         fallback_title=self.item.title,
                     ),
-                    "image": season_image,
-                },
-            )
+                    image=season_image,
+                )
```
Comment (ryan-winkler) [inline src/app/models/tv.py]: This block needs both an allowed-bucket rule and the existing database-safe create behavior. `find_item_across_buckets()` can fall back to an anime candidate for a normal TV parent when the preferred bucket is absent, and the following plain `create()` can race after two requests both observe no item. The replacement below prefers only compatible buckets and uses `get_or_create()` for the target bucket.

```suggestion
            target_bucket = (
                MediaTypes.ANIME.value
                if self.item.library_media_type == MediaTypes.ANIME.value
                else MediaTypes.SEASON.value
            )
            season_identity = {
                "media_id": self.item.media_id,
                "source": self.item.source,
                "media_type": MediaTypes.SEASON.value,
                "season_number": season_number,
            }
            allowed_buckets = [self.item.library_media_type]
            if target_bucket not in allowed_buckets:
                allowed_buckets.append(target_bucket)

            item = None
            for bucket in allowed_buckets:
                item = (
                    Item.objects.filter(
                        **season_identity,
                        library_media_type=bucket,
                    )
                    .order_by("id")
                    .first()
                )
                if item is not None:
                    break

            if item is None:
                item, _ = Item.objects.get_or_create(
                    **season_identity,
                    library_media_type=target_bucket,
                    defaults={
                        **Item.title_fields_from_metadata(
                            season_metadata,
                            fallback_title=self.item.title,
                        ),
                        "image": season_image,
                    },
                )
```

File: src/app/models/tv.py [hunk 2]
```diff
@@ -1008,8 +1024,21 @@ def _sync_status_after_episode_change(self):
...
-            # No release data, but we have watches — stay in progress
-            desired_status = Status.IN_PROGRESS.value
+            # No release events for this season. That is not evidence the
+            # season is unfinished - events are only populated once the
+            # calendar task has run for this item, and imported libraries
+            # often have none at all. Fall back to the media-server episode
+            # count when we have one; only assume "still watching" when we
+            # genuinely have no episode total from any local source.
+            local_total = getattr(self.item, "local_season_episode_count", None)
+            if local_total and max_watched >= local_total:
```
Comment (ryan-winkler) [inline src/app/models/tv.py]: `max_watched` is the highest episode position. It is not the number of completed episodes. With `local_total=10`, a single watch of episode 10 satisfies this condition and marks the season complete. The model already provides a distinct completed-episode count that ignores repeated plays and non-completed rows.

```suggestion
            if local_total and self.completed_episode_count >= local_total:
```

Please cover the sparse E10-only case, duplicate plays, full distinct completion, and the manual `In progress` rewatch override.

**Review by ryan-winkler — CHANGES_REQUESTED** [review]
The reported failure is real and the PR is aimed at the correct user problem. The concrete imported-library evidence in the description is especially useful. I found three blocking correctness points that need to be resolved before this can merge:

1. `max_watched` is an episode position, not a completed-episode count. Sparse history can therefore become `Completed` incorrectly.
2. The new lookup-then-`create()` path removes the database-safe `get_or_create()` behavior and can fail under concurrent completion requests.
3. The unbounded cross-bucket fallback can select an anime season for a normal TV parent, or the reverse, when the preferred bucket is absent.

I added exact code suggestions for the two changed paths. Equivalent code is fine if it preserves these rules. I also created #812 as the regression contract and enriched this PR with the full test matrix, relationships, `/qa` result, interaction-scope statement, and post-mortem.

Please add focused tests for:

- reuse of an imported `tv`-bucket season without creating a second `Item`, `Season`, or episode set;
- preservation of the existing episode dates;
- exclusion of anime candidates for normal TV and normal-TV candidates for anime;
- one late episode out of ten remaining `In progress`;
- all ten distinct completed episodes becoming `Completed`;
- repeated plays not increasing the distinct completion count;
- a deliberate `In progress` rewatch remaining `In progress`.

The branch also needs an update from current `latest` before validation. Record the targeted tests, fast suite, Ruff, migration dry-run, and diff check after that update. The AI disclosure also needs the exact provider and model identifier; the current wording is not specific enough to verify.

No screenshot is required for this backend-only change. A manual data check should still confirm that old dates remain visible after whole-show completion.

Thank you for isolating the bucket mismatch and for opening a draft rather than presenting this as merge-ready. The issue is worth fixing; these changes will make the repair safe for imported libraries and future identity variants.
